src/Tools/dotnet-monitor/CollectionRules/ActionListExecutor.cs

```diff
@@ -33,47 +35,100 @@ public ActionListExecutor(ILogger<CollectionRuleService> logger, ICollectionRule
             }
 
             int actionIndex = 0;
+            List<ActionCompletionEntry> deferredCompletions = new(context.Options.Actions.Count);
 
-            foreach (CollectionRuleActionOptions actionOption in context.Options.Actions)
+            try
             {
-                // TODO: Not currently accounting for properties from previous executed actions
-
-                KeyValueLogScope actionScope = new();
-                actionScope.AddCollectionRuleAction(actionOption.Type, actionIndex);
-                using IDisposable actionScopeRegistration = _logger.BeginScope(actionScope);
+                // Start and optionally wait for each action to complete
+                foreach (CollectionRuleActionOptions actionOption in context.Options.Actions)
+                {
+                    // TODO: Not currently accounting for properties from previous executed actions
 
-                _logger.CollectionRuleActionStarted(context.Name, actionOption.Type);
+                    KeyValueLogScope actionScope = new();
+                    actionScope.AddCollectionRuleAction(actionOption.Type, actionIndex);
+                    using IDisposable actionScopeRegistration = _logger.BeginScope(actionScope);
 
-                try
-                {
-                    ICollectionRuleActionFactoryProxy factory;
+                    _logger.CollectionRuleActionStarted(context.Name, actionOption.Type);
 
-                    if (!_actionOperations.TryCreateFactory(actionOption.Type, out factory))
+                    try
+                    {
+                        ICollectionRuleActionFactoryProxy factory;
+
+                        if (!_actionOperations.TryCreateFactory(actionOption.Type, out factory))
+                        {
+                            throw new InvalidOperationException(Strings.ErrorMessage_CouldNotMapToAction);
+                        }
+
+                        ICollectionRuleAction action = factory.Create(context.EndpointInfo, actionOption.Settings);
+
+                        try
+                        {
+                            await action.StartAsync(cancellationToken);
+
+                            // Check if the action completion should be awaited synchronously (in respected to
+                            // starting the next action). If not, add a deferred entry so that it can be completed
+                            // add the end of the execution list.
```

Comment:
I'm guessing this should be "at" or "after" the end of the execution list, but I may be misinterpreting the comment

Reply:
Either way, but yet, you are correct; I'll go with "after".